bindings/java/src/main/com/apple/foundationdb/DirectBufferPool.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * DirectBufferPool.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2020 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb;
+
+import java.io.Closeable;
+import java.nio.ByteBuffer;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * A singleton that manages a pool of {@link DirectByteBuffer}, that will be
+ * shared by the {@link DirectBufferIterator} instances. It is responsibilty of
+ * user to return the borrowed buffers.
+ */
+class DirectBufferPool {
+	static final DirectBufferPool __instance = new DirectBufferPool();
+
+	// When tuning this, make sure that the size of the buffer,
+	// is always greater than the maximum size KV allowed by FDB.
+	static private final int DEFAULT_NUM_BUFFERS = 128;
+	static private final int DEFAULT_BUFFER_SIZE = 1024 * 512;
+
+	private ArrayBlockingQueue<ByteBuffer> buffers;
+
+	public DirectBufferPool() {
+		resize(DEFAULT_NUM_BUFFERS, DEFAULT_BUFFER_SIZE);
+	}
+
+	public static DirectBufferPool getInstance() {
+		return __instance;
+	}
+
+	// Resizes buffer pool given number and size. Throws OutOfMemory exception
+	// if unable to allocate as asked.
+	public synchronized void resize(int newSize, int bufferSize) {
+		buffers = new ArrayBlockingQueue<>(newSize);
+		while (buffers.size() < newSize) {
+			ByteBuffer buffer = ByteBuffer.allocateDirect(bufferSize);
+			buffers.add(buffer);
+		}
+	}
+
+	/**
+	 * Requests a {@link #DirectByteBuffer} from our pool. Returns null if pool is empty.
+	 */
+	public synchronized ByteBuffer poll() {
+		return buffers.poll();
+	}
+
+	/**
+	 * Returns the {@link #DirectByteBuffer} that was borrowed from our pool. This
+	 * is non-blocking as it was borrowed from this pool.
```

Comment:
I think, if I'm reading the contract of `offer` correctly, this is non-blocking because it will drop the offered `ByteBuffer` if the queue is full.

Reply:
I missed responding to first comment. You are correct, it will reject if full. The user of this class is responsible to use it correctly, so it should be accepted in all cases as it will return what was taken. Except for the case of resize, in which case we will reject the buffers which are not of current size we wish to keep.

I think "#" was the problem. I've removed that, and at-least on my Emacs setup it follows. (I guess there must be some way to check broken links in javadocs which perhaps we should add to our CI. Will look into it.)

Reply:
Sorry, the point I was trying to make is that I think the Javadoc comment here is incorrect, because the function will be non-blocking, but not because the buffer was borrowed from the pool (as the comment suggests). It might make sense to either remove that sentence or remove the "as" clause (to avoid confusing future readers).